asic/src/tofino_asic/ports.rs

```diff
@@ -412,6 +413,45 @@ pub fn set_prbs(
     Ok(())
 }
 
+const MAX_N_LANES: usize = 16;
+const fn empty_ber() -> bf_port_ber_t {
+    bf_port_ber_t {
+        ber: bf_port_ber_t_ber_union_ {
+            tof2: bf_tof2_ber_t {
+                ber_aggr: 0.0,
+                ber_per_lane: [0.0; MAX_N_LANES],
+                sym_err_ctrs: [0; MAX_N_LANES],
+            },
+        },
+    }
+}
+
+/// Return the bit-error for a port.
+pub fn get_ber(hdl: &Handle, port_hdl: PortHdl) -> AsicResult<Ber> {
+    let dev = hdl.dev_id;
+    let mut fp = FrontPortHandle::from_port_hdl(hdl, port_hdl)?;
+    let mut ber = empty_ber();
+    let ber = match unsafe {
+        bf_pm_port_ber_get(dev, fp.ptr(), &mut ber as *mut _)
+    } {
+        BF_SUCCESS => {
+            // Safety: We're using the `empty_ber()` function above to
+            // initialize this, so we know this field exists and is the active
+            // union field. It's also the only one.
+            unsafe { ber.ber.tof2 }
+        }
+        e => {
+            slog::error!(hdl.log, "failed to get port BER: {:?}", e);
+            return Err(sde_error("getting port BER", e));
+        }
+    };
+    Ok(Ber {
+        symbol_errors: ber.sym_err_ctrs.to_vec(),
+        ber: ber.ber_per_lane.to_vec(),
+        total_ber: ber.ber_aggr,
```

Comment:
I'd prefer to extract only the counters that are specific to the lanes this link uses, but the SDE does not make it easy to understand which those are. It relies on some bit-twiddling that's done [here](https://github.com/oxidecomputer/tofino-sde/blob/0e4cfe10c92db223a953923d77f09a299b82157b/pkgsrc/bf-drivers/src/bf_pm/pm.c#L2120-L2140) -- it's not super complicated, but it also feels a bit fraught to reimplement this in the ASIC layer. I can do that if we think it's valuable, though.

Reply:
I added this in https://github.com/oxidecomputer/dendrite/pull/20/commits/f170ce7fdcfb1aaed02ac8d6394c4672e331bdfb. I'm ambivalent about it, since we need to make additional calls into the SDE to get the speed of the ASIC port, and then duplicate some opaque logic that's in the SDE to figure out the right counters. But it's much more helpful to report only the counters that _should_ matter for a link, so I've included it for now.